curriculum/challenges/english/25-front-end-development/lab-hash-table/67ed03ac474c48692f41749e.md

```diff
@@ -0,0 +1,360 @@
+    len(sig.parameters) == 2
+  `))
+})
+```
+
+The `HashTable` class should have a `remove` method.
+
+```js
+({test: () => {assert(runPython(`_Node(_code).find_class('HashTable').has_function('remove')`))}})
+```
+
+The `remove` method should take a key as a parameter.
+
+```js
+({ test: () => assert(runPython(`
+    import inspect 
+    ht= HashTable()
+    sig = inspect.signature(ht.remove)
+    len(sig.parameters) == 1
+  `))
+})
+```
+
+The `HashTable` class should have a `lookup` method.
+
+```js
+({test: () => {assert(runPython(`_Node(_code).find_class('HashTable').has_function('lookup')`))}})
+```
+
+The `lookup` method should take a key as the parameter.
+
+```js
+({ test: () => assert(runPython(`
+    import inspect 
+    ht= HashTable()
+    sig = inspect.signature(ht.lookup)
+    len(sig.parameters) == 1
+  `))
+})
+```
+
+`HashTable.hash("golf")` should return `424`.
+
+```js
+({ 
+    test: () => runPython(`
+        ht = HashTable()
+        hash_result = ht.hash("golf")
+        assert hash_result == 424
+
+        # prevent hardcoding
+        assert ht.hash("dear") == 412
+        assert ht.hash("cat") == 312
+    `) 
+})
+```
+
+`HashTable.add("golf", "sport")` should add the key-value pair to the collection at key `424`.
+
+```js
+
+({ 
+    test: () => runPython(`
+    ht = HashTable()
+    ht.add("golf", "sport")
+    expected_value = {424: {'golf': 'sport'}}
+
+    assert ht.collection == expected_value
+    `) 
+})
+```
+
+`HashTable.add("dear", "friend")` and `HashTable.add("read", "book")` should add both the key-value pairs to the collection at index `412` as a nested dictionary.
+
+```js
+({ 
+    test: () => runPython(`
+ 
+    ht = HashTable()
+
+    ht.add("dear", "friend")
+    ht.add("read", "book")
+
+    expected_value = {
+        "dear": "friend",
+        "read": "book"
+    }
+
+    assert ht.collection.get(412) == expected_value
+    `) 
+})
+```
+
+When the `"golf", "sport"` key-value pair exists in the hash table, `HashTable.remove("golf")` should remove the key-value pair from the collection.
+
+```js
+({ 
+    test: () => runPython(`
+    ht = HashTable()
+    ht.add("golf", "sport")
+
+    expected_value_before_removal = {
+        "golf": "sport"
+    }
+
+    index = ht.hash("golf")
+
+    assert ht.collection.get(index) == expected_value_before_removal
+
+    ht.remove("golf")
+
+    assert "golf" not in ht.collection.get(index, {})
+    `) 
+})
+```
+
+When the `"golf", "sport"` key-value pair exists in the hash table, `HashTable.lookup("golf")` should return `"sport"`.
+
+```js
+({ 
+    test: () => runPython(`
+    ht = HashTable()
+
+    ht.add("golf", "sport")
+
+    expected_value = "sport"
+    assert ht.lookup("golf") == expected_value
+    `) 
+})
+
+```
+
+When the `"golf", "sport"` key-value pair does not exist in the collection, `HashTable.lookup("golf")` should return `None`.
+
+```js
+({ 
+    test: () => runPython(`
+        ht = HashTable()
+        ht.remove("golf")
+
```

Comment:
```suggestion
```
This can be removed because `ht` is a new instance, so its `collection` should be empty.

Because of this I found something we need to test. I've written a `remove` method that throws when the key to remove is not in the collection and this test was failing, but none of the tests for `remove`.

So I think we should require that when the key does not exists in the collection, `remove` should not raise an exception. What do you think?

Reply:
I have modified the instruction for `remove` and also added a few more tests.